Refactor ControllerRefManager

enisoc opened the pull request:
**What this PR does / why we need it**:

To prepare for implementing ControllerRef across all controllers (https://github.com/kubernetes/community/pull/298), this pushes the common adopt/orphan logic into ControllerRefManager so each controller doesn't have to duplicate it.

This also shares the adopt/orphan logic between Pods and ReplicaSets, so it lives in only one place.

**Which issue this PR fixes**:

**Special notes for your reviewer**:

**Release note**:
```release-note
```

cc @kubernetes/sig-apps-pr-reviews

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41361)
<!-- Reviewable:end -->

kargakis commented on pkg/controller/controller_ref_manager.go:
```diff
@@ -25,106 +25,155 @@ import (
+			return true, nil
+		}
+		// Owned by us but selector doesn't match.
+		if m.controller.GetDeletionTimestamp() != nil {
```
Move this on top of the method?

kow3ns commented:
Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks broke.

---

*[pkg/controller/controller_ref_manager.go, line 58 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41361#-KcyQvjaaDgETxrkcLSD:-KcyQvjqROoBFPtAEw52:b-a6n7xg) ([raw file](https://github.com/kubernetes/kubernetes/blob/44415dc341e7546e084fc1ae3a5c66141187ff11/pkg/controller/controller_ref_manager.go#L58)):*
> ```Go
> 			return false, nil
> 		}
> 		if m.selector.Matches(labels.Set(obj.GetLabels())) {
> ```

I'd compute 
```golang
matches := m.selector.Matches(labels.Set(obj.GetLabels())) 
deleted := m.controller.GetDeletionTimestamp() != nil
```
and use the computed values throughout the rest of the function

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41361)*
<!-- Sent from Reviewable.io -->

Commit:
Refactor ControllerRefManager

To prepare for implementing ControllerRef across all controllers,
this pushes the common adopt/orphan logic into ControllerRefManager
so each controller doesn't have to duplicate it.

This also shares the adopt/orphan logic between Pods and ReplicaSets,
so it lives in only one place.

enisoc commented:
Review status: 5 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/controller/controller_ref_manager.go, line 58 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41361#-KcyQvjaaDgETxrkcLSD:-Kd6wc5uhfRrzOT8-O92:bnlcm1l) ([raw file](https://github.com/kubernetes/kubernetes/blob/44415dc341e7546e084fc1ae3a5c66141187ff11/pkg/controller/controller_ref_manager.go#L58)):*
<details><summary><i>Previously, kow3ns (Kenneth Owens) wrote…</i></summary><blockquote>

I'd compute 
```golang
matches := m.selector.Matches(labels.Set(obj.GetLabels())) 
deleted := m.controller.GetDeletionTimestamp() != nil
```
and use the computed values throughout the rest of the function
</blockquote></details>

This function will get called for every single Pod in the cluster, multiplied by the number of controllers. So I want to be careful about doing as little work as possible for the majority case of Pods we don't care about.

In the New World, almost all Pods should have ControllerRefs almost all the time. Of those, the vast majority will end up being owned by someone other than this particular controller. In that case, both `matches` and `deleted` are unneeded so we should short-circuit the computation.

---

*[pkg/controller/controller_ref_manager.go, line 63 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41361#-KctdqRKSzBmqEGa_YwX-r1-63:-Kd7-5ayPNePshd2EPBL:b-ri53er) ([raw file](https://github.com/kubernetes/kubernetes/blob/44415dc341e7546e084fc1ae3a5c66141187ff11/pkg/controller/controller_ref_manager.go#L63)):*
<details><summary><i>Previously, kargakis (Michail Kargakis) wrote…</i></summary><blockquote>

Move this on top of the method? 
</blockquote></details>

The deletion timestamp should prevent us from taking action, but it doesn't mean we should always return `false` (not claimed). If we already own the object, claiming it is a no-op and thus is allowed during deletion. This is important so the controller continues to count objects it already owns when updating its status during deletion.

 I've edited the comments to describe this.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41361)*
<!-- Sent from Reviewable.io -->

enisoc commented:
@k8s-bot cvm gke e2e test this

kow3ns commented:
Review status: 5 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/controller/controller_ref_manager.go, line 58 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41361#-KcyQvjaaDgETxrkcLSD:-Kd7n5tOWKOi3iR3Hggo:bwqlh9c) ([raw file](https://github.com/kubernetes/kubernetes/blob/44415dc341e7546e084fc1ae3a5c66141187ff11/pkg/controller/controller_ref_manager.go#L58)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary><blockquote>

This function will get called for every single Pod in the cluster, multiplied by the number of controllers. So I want to be careful about doing as little work as possible for the majority case of Pods we don't care about.

In the New World, almost all Pods should have ControllerRefs almost all the time. Of those, the vast majority will end up being owned by someone other than this particular controller. In that case, both `matches` and `deleted` are unneeded so we should short-circuit the computation.
</blockquote></details>

Makes sense, but if almost all Pods have ControllerRefs almost all of  the time than the expected branch is 

```golang
if controllerRef != nil {
```
you are expecting to invoke 

```golang
m.selector.Matches(labels.Set(obj.GetLabels()))
```
so you should just compute it up front and then use its inverse in the less likely case below. That should not hurt performance, but, given your prediction of the likely branch, which is reasonable, it won't improve it much either.  I was suggesting that it would add clarity.

```golang
m.contoller.GetDeleteionTimestamp()
``` 
Should not be expensive in either case. Again, this was a suggestion for clarity.

If both cases the the comments are good enough that the intention is  captured.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41361)*
<!-- Sent from Reviewable.io -->

kow3ns commented:
Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41361)*
<!-- Sent from Reviewable.io -->

enisoc commented:
Review status: 8 of 9 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/controller/controller_ref_manager.go, line 58 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41361#-KcyQvjaaDgETxrkcLSD:-Kd84fJFJXKIOU_yKLVx:bvkcmtk) ([raw file](https://github.com/kubernetes/kubernetes/blob/44415dc341e7546e084fc1ae3a5c66141187ff11/pkg/controller/controller_ref_manager.go#L58)):*
> The expected branch is `if controllerRef != nil {`

That's true, but then within that branch, the most likely sub-branch by far is:

```go
if controllerRef.UID != m.controller.GetUID() {
	// Owned by someone else. Ignore.
	return false, nil
}
```

That's because any single controller probably owns a small percentage of all Pods in the cluster. In this case, I don't want to pay for checking the label selector.

In reality it's probably not a big difference in performance, but I think the code is clear both ways, so I don't see a benefit to changing it.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41361)*
<!-- Sent from Reviewable.io -->

kow3ns commented:
/lgtm /approve

---







*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41361#-:-KdCJXCt0PHbD46cQ2_U:b-bdkrmo)*
<!-- Sent from Reviewable.io -->

kow3ns reviewed (approved):
/lgtm /approve

enisoc commented:
@janetkuo Can you /approve for OWNERS?

janetkuo commented:
/approve

kow3ns commented:
/lgtm

enisoc commented:
@k8s-bot test this

enisoc commented:
@k8s-bot test this

enisoc commented:
@k8s-bot gce etcd3 e2e test this
@k8s-bot cvm gce e2e test this

enisoc commented:
@k8s-bot gce etcd3 e2e test this
@k8s-bot cvm gce e2e test this

enisoc commented:
@k8s-bot kops aws e2e test this

enisoc commented:
@k8s-bot kubemark e2e test this
